### PR TITLE
Andy/prolly indexes

### PR DESCRIPTION
**Pull request by andrew-wm-arthur**
This PR adds a second implementation of `durable.Index` and attempts to refactor pkg `sqle` to use `durable.Index` in place of `types.Map`. This is a major refactor in the SQL engine with the aim of supporting both the existing storage format and the new-and-improved™ storage format. As a measure of progress roughly 70% of the engine tests are passing against `sqle` while using the new format. 

Various changes were made in pkg `val` in order to support the SQL types/encodings need for sysbench. Some of these encodings are experimental and will change in the future.

### CODE REVIEW

**Comment by andrew-wm-arthur** [comment]
#benchmark

**Comment by andrew-wm-arthur** [comment]
#benchmark

File: go/libraries/doltcore/sqle/writer/prolly_table_writer.go [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+	if err := m.mut.Delete(ctx, oldKey); err != nil {
+		return err
+	}
+
+	for to, from := range m.keyMap {
+		m.keyBld.PutField(to, newRow[from])
+	}
+	newKey := m.keyBld.Build(sharePool)
+
+	ok, err := m.mut.Has(ctx, newKey)
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/writer/prolly_table_writer.go:241]: This tracks unapplied mutations?

File: go/libraries/doltcore/sqle/writer/prolly_table_writer.go [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+		}
+		return
+	})
+
+	s := m.keyBld.Desc.Format(key)
+
+	return sql.NewUniqueKeyErr(s, true, existing)
+}
+
+type colMapping []int
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/writer/prolly_table_writer.go:277]: could use a comment

File: go/store/prolly/node_store.go [hunk 3]
```diff
@@ -33,36 +41,235 @@ type NodeStore interface {
...
 }
+
+func (ns nodeStore) Format() *types.NomsBinFormat {
+	// todo(andy): read from |ns.store|
+	return types.Format_DOLT_1
+}
+
+type centry struct {
+	c    chunks.Chunk
+	i    int
```
Comment (zachmu) [inline go/store/prolly/node_store.go:107]: ?

File: go/store/val/codec.go [hunk 4]
```diff
@@ -55,27 +61,26 @@ type Encoding uint8
...
-	Int16Enc   Encoding = 3
-	Uint16Enc  Encoding = 4
-	Int24Enc   Encoding = 5
-	Uint24Enc  Encoding = 6
+	NullEnc   Encoding = 0
+	Int8Enc   Encoding = 1
+	Uint8Enc  Encoding = 2
+	Int16Enc  Encoding = 3
+	Uint16Enc Encoding = 4
+	// Int24Enc   Encoding = 5
```
Comment (zachmu) [inline go/store/val/codec.go:69]: ?

**Review by zachmu — APPROVED** [review]
Overall looks good, just a few comments.

Seems like the new prolly store files could use some unit tests, e.g. node_store_test.go. Probably a good idea to add those as soon as the interface stabilizes.

File: go/libraries/doltcore/sqle/writer/prolly_table_writer.go (hunk 1 above)
Comment (andrew-wm-arthur) [reply go/libraries/doltcore/sqle/writer/prolly_table_writer.go:241]: yeah buddy, got a sweet skip list in there

File: go/store/prolly/node_store.go (hunk 3 above)
Comment (andrew-wm-arthur) [reply go/store/prolly/node_store.go:107]: "cache entry" stole this from Aaron's implementation in `dolthubapi`

File: go/store/val/codec.go (hunk 4 above)
Comment (andrew-wm-arthur) [reply go/store/val/codec.go:69]: still undecided if we're going to support a 3 byte integer encoding

**Pull request merged**